includes/payment-methods/class-wc-stripe-payment-request.php

```diff
@@ -1460,20 +1393,26 @@ protected function build_shipping_methods( $shipping_methods ) {
 		return $shipping;
 	}
 
+
 	/**
-	 * Builds the line items to pass to Payment Request
+	 * Builds response to pass to the Payment Request
 	 *
 	 * @since   3.1.0
 	 * @version 4.0.0
+	 *
+	 * @param bool $itemized_display_items Wether to return an array of items with its details or not.
+	 * @param bool $has_shipping_address True is the user has picked or entered a shipping address on the payment dialog.
 	 */
-	protected function build_display_items( $itemized_display_items = false ) {
+	protected function build_response( $itemized_display_items = false, $has_shipping_address = false ) {
```

Comment:
@ricardo Unfortunately, `$has_shipping_address` is needed here. We talked about using Cart' session data to know if the Cart has a Shipping address attached to it but this is not always desirable.

A shipping address may be present on the session even if the user hasn't opened the payment dialog. So when the user clicks the PR button on the product page for the very first time the server will return an amount for "Shipping". That's why we need to pass the state of "has_shipping_address" from the frontend to respond accordingly.

Reply:
@asumaran I found an issue with excluding tax based on `has_shipping_address` and I guess we had this issue before.

Tax is not always based on shipping address. It can be based on store address or billing address and thus be applied to virtual products that don't require shipping.

Steps to reproduce:
1. Go to **WooCommerce > Settings** and Enable Taxes.
2. Go to **WooCommerce > Settings > Taxes**:
  i. Check "No, I will enter prices exclusive of tax".
  ii. Calculate tax based on: "Shop base address".
  iii. Shipping tax class: "Standard".
3. Go to the Album product page and try to pay with PRB.
4. Notice the "Tax" label is always `0` because of `has_shipping_address`.

Reply:
> A shipping address may be present on the session even if the user hasn't opened the payment dialog.

I'm starting to think it's fine to allow the session address to be pre-selected and/or display a shipping label with a default value.

It will be something already familiar to the user and it's the same experience as if the user clicks the PRB from the cart page, right?

Another point to consider is that it might be confusing to display "Shipping: $0" when in reality there will be a shipping cost after address selection. One way we could prevent this would be by only adding the shipping label if `has_shipping_address` is `true`. Like so:

```php
if ( $has_shipping_address ) {
	$items[] = [
		'label'   => esc_html( __( 'Shipping', 'woocommerce-gateway-stripe' ) ),
		'amount'  => WC_Stripe_Helper::get_stripe_amount( $shipping ),
		'pending' => true,
	];
}
```

But again, maybe it's ok to display the session shipping value on the product page.

Another idea would be cleaning up the session shipping address upon clicking the PRB with something like `WC()->customer->set_shipping_location( ... )` (haven't tested though). But perhaps not the best UX.

What do you think?

Reply:
>  I found an issue with excluding tax based on has_shipping_address and I guess we had this issue before.
> Tax is not always based on shipping address. It can be based on store address or billing address and thus be applied to virtual products that don't require shipping.

@ricardo Could you elaborate on this? I don't have the Album product on my instance anymore but I've tried with another virtual product. Tax is `0` on the product page and even on the cart and checkout pages. So, I'm not sure how much the tax amount should be.

BTW I have another instance running simultaneously with the `trunk` branch so I can verify how it worked before my changes. In both my branch and trunk the behavior is currently the same with the steps you mention.  Please add more steps or even screenshots so I can verify and work on a solution.

Reply:
> I'm starting to think it's fine to allow the session address to be pre-selected and/or display a shipping label with a default value.

I was feeling the same but I was thinking it's more important to not introduce new behaviors on this issue given that the issue is mostly related to refactor how totals were obtained/calculated on the product page. I constantly checked with trunk how things worked before introducing any changes.

Let me know if this is small enough to be introduced on this PR. I'll be happy to work on it.

Reply:
> Could you elaborate on this?

Sorry, I forgot to add that you need to go to **WooCommerce > Settings > Tax** and under Shipping classes, add a class of 10% (Edited above). The tax on the Payment Request shouldn't be `0`.

If you go to the cart page, right above the PRB, it displays the tax correctly, but not on the Payment Request labels on the product page.

As I mentioned, apparently this is an issue from before, but I think it should be fixed in this PR, because the issue is actually not fixed because of [this check](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/e7e17c3b3f648c46bc8acddf8aaca27d777816a1/includes/payment-methods/class-wc-stripe-payment-request.php#L1414-L1417) for a shipping address. Virtual products won't have shipping addresses and therefore we can't display tax labels to them.

Reply:
@ricardo I'll take a look

Reply:
@ricardo I've tried to reproduce the issue you mention without success.

I've created a [simple](https://user-images.githubusercontent.com/1025173/121595993-7fef4800-ca04-11eb-8836-3c3706885e86.png) virtual product. Then created a [10% standard tax rate](https://user-images.githubusercontent.com/1025173/121596185-c0e75c80-ca04-11eb-8878-b35c4d37d438.png) and I can see the [tax being applied ](https://user-images.githubusercontent.com/1025173/121596306-e4120c00-ca04-11eb-95bf-96fd26d4c8ea.png). Is the issue already fixed on my branch or I am missing something?